# 2023년 4월 30일(일)

---

### Today I Learned 

- [iOS-Swift] 에러처리(Error Handling)

---

### 에러 처리의 과정(3단계)

1. 에러 정의(어떤 에러가 발생할지 경우를 미리 정의)
2. 에러가 발생할 수 있는 함수에 대한 정의
   - 에러가 발생 가능한 상황에 throw 키워드로 에러를 던질(throwing)수 있도록 정의
3. 에러가 발생할 수 있는 함수의 처리(함수의 실행) => 기존과 조금 다르다. (try와 do-catch 문으로 처리)
   - 정상적인 처리 방법: do-catch 블럭에서 처리(조금 간편한 방식도 존재)

```swift
// 1) 에러 정의(어떤 에러가 발생할지 경우를 미리 정의)
enum HeightError: Error { // 에러 프로토콜 채택 (약속)
case maxHeight
case minHeight
}

// 2) 에러가 발생할 수 있는 함수에 대한 정의
// 에러 발생 가능한 상황에 throw 키워드로 에러를 던질(throwing)수 있도록 정의
func checkingHeight(height: Int) throws -> Bool { // (에러를 던질 수 있는 함수 타입이다.)
    if height > 190 {
        throw HeightError.maxHeight
    } else if height < 130 {
        throw HeightError.minHeight
    } else {
        return height >= 160 ? true : false
    }
}

// 3) 에러가 발생할 수 있는 함수의 처리(함수의 실행) ==> 기존과 조금 다르다. (try와 do-catch 문으로 처리)
// 정상적인 처리방법: do-catch 블럭에서 처리(조금 간편한 방식도 존재)
// do 블럭 - 함수를 통한 정상적인 처리의 경우 실행하는 블럭
// catch 블럭 - 함수가 에러를 던졌을 경우의 처리 실행하는 블럭
do {
    let isChecked = try checkingHeight(height: 200)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch {
    print("놀이기구 타는 것 불가능")
}
```

<br/>

### 에러를 처리하는 방법 - try / try? / try!

1. try
   - 모든 에러 발생의 예외적인 경우를 디테일하게 처리 가능 
2. try? (Optional try)
   - 정상적인 경우 => (함수의) 리턴 타입 
   - 에러발생하면 => nil
   - 결국 옵셔널 타입으로 리턴하기 때문에 벗겨서(unwrapping) 사용해야 함 
3. try! (Forced try)
   - 정상적인 경우 => (함수의) 리턴 타입
   - 에러 발생하면 => 에러 발생 
   - 에러가 발생할 가능성이 없는 경우에 제한적으로 사용 

```swift
/**===========================================
 1) 에러 정식 처리 방법
- 모든 에러발생의 예외적인 경우를 디테일하게 처리 가능
============================================**/
do {
    let isChecked = try checkingHeight(height: 160)
    
    isChecked ? print("청룡열차 가능") : print("후룸라이드 가능")
} catch {
    print("놀이기구 타는 것 불가능")
}

/**===========================================
 2) try? (옵셔널 트라이) ===> 옵셔널 타입으로 리턴

   (1) 정상적인 경우 ==> (정상)리턴타입으로 리턴 ==> 옵셔널 타입이 감싸진 형태로 리턴되므로 벗겨서(unwrapping)해서 사용해야 함
   (2) 에러가 발생하면 ==> nil 리턴
============================================**/
let isChecked = try? checkingHeight(height: 160) // Bool? --> 당연히, 옵셔널 타입을 벗겨서 사용해야함
print(isChecked) // Optional(true)

/**===========================================
 3) try! (Forced 트라이) ===> 에러가 날 수 없는 경우에만 사용 가능

   (1) 정상적인 경우 ==> (정상)리턴타입으로 리턴 => 옵셔널 타입 X
   (2) 에러가 발생하면 ==> 런타임에러
============================================**/
let isChecked2 = try! checkingHeight(height: 160) // Bool --> 에러가 발생할 수 없다고 확신이 있는 경우만 사용해야 함.
print(isChecked2) // true
```

<br/>

### Catch 블럭 처리방법

1. 패턴을 모두 정의(경우를 모두 따로 처리)
   - 각 Catch 블럭에서 모든 경우를 정의하거나, 구체적인 경우를 먼저 정의하고 마지막엔 catch 키워드만 있어도 됨 
2. 패턴없이 에러를 받아서 처리 
   - catch 블럭 내에서 error 상수 제공 
   - error 상수(에러 프로토콜 타입)를 구체적으로 정의한 타입으로 캐스팅해서 Switch문 등으로 다시 처리 

```swift
// catch블럭은 do블럭에서 발생한 에러만을 처리하는 블럭
// 모든 에러를 반드시 처리해야만 함 (글로벌 스코프에서는 모든 에러를 처리하지 않아도 컴파일 에러발생하지 않음)

// 패턴이 있는 경우(모든 에러를 각각 따로 처리 해야함)
do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch HeightError.maxHeight  {    // where절을 추가해서, 매칭시킬 에러패턴에 조건을 추가할 수 있음
    print("키가 커서 놀이기구 타는 것 불가능")
} catch HeightError.minHeight {      // 생략가능
    print("키가 작아서 놀이기구 타는 것 불가능")
}

// catch 패턴이 없이 처리도 가능
do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch {    // error 상수를 제공 (모든 에러가 넘어옴)
    if let error = error as? HeightError {    // 실제 우리가 정의한 구체적인 에러 타입이 아니고, 에러 타입(프로토콜)이 넘어올 뿐
        switch error {
        case .maxHeight:
            print("키가 커서 놀이기구 타는 것 불가능")
        case .minHeight:
            print("키가 작아서 놀이기구 타는 것 불가능")
        }
    }
}

// Swift 5.3이후 catch블럭의 처리
do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch HeightError.maxHeight, HeightError.minHeight {   // 케이스 나열도 가능해짐
    print("놀이기구 타는 것 불가능")
}
```

